### DISCUSSION TITLE
How does one handle the ```$container``` property while manually rewriting the AST?

### DISCUSSION THREAD

**Discussion by soft-guy** [eclipse-langium/langium#1767] [Q&A] [answered]
## Context:
Some parser rules of the grammar of a language I am developing represent syntactic sugar. Instead of writing additional generation code for the AST nodes resulting from these rules, I want to rewrite the AST to eliminate the syntactic sugar.

---
## Workaround
To achieve this, I need to create a new object of a specific type whose properties are to reference objects accessible through the properties of the original syntactic sugar node. After creating this new object, I would substitute it for the syntactic sugar node in the AST. However, I have some challenges with the ```$container``` property.

---
### 1. The Problem with ```$container```:
- I can assign ```$container``` property of a new object while creating it.
- However, the objects used to construct this new object retain their ```$container``` property values, which are now incorrect since they refer to the old syntactic sugar node, which is no longer in the AST.
- Since the ```$container``` property is read-only, I cannot directly update it to reflect the new AST structure.

---
### 2. Possible Workarounds:
- Creating new objects that replicate the properties of the existing nodes but assign the correct ```$container``` values manually while creating.
- However, this approach requires recursively updating all related nodes down to the leaves, which seems cumbersome and inelegant.
- Alternatively, I could ignore handling ```$container``` property, at least occasionally, but this might cause issues in other parts that depend on its correctness.

---
## Question:
Is there an established, recommended, or conventional way in Langium to handle ```$container``` updates when manually modifying the AST? Alternatively, how critical is the ```$container``` property for AST consistency, and can it be safely ignored in some scenarios?

**Comment by msujew** [comment, marked as answer]
Hey @soft-guy,

we have a utility function [`AstUtils.linkContentToContainer`](https://eclipse-langium.github.io/langium/functions/langium.AstUtils.linkContentToContainer.html) that you can use for this exact purpose. During the AST construction, you can just ignore the `$container` property (for example, by setting it to `undefined!`), and then use the linked function to let the framework take over the assignment.
